politeiad/backendv2/tstorebe/tstorebe.go

```diff
@@ -1017,21 +1019,76 @@ func (t *tstoreBackend) Fsck() error {
 		return err
 	}
 
-	// Rebuild the inventory cache. This is a multi-step process:
-	//
-	// 1. Sort the tokens into two groups; unvetted and vetted. Each
-	//    group is additionally sorted by the timestamp of their most
-	//    recent status change.
-	//
-	// 2. Add the vetted tokens to the inventory cache. They MUST be
-	//    added starting with the oldest timestamp. The vetted token
-	//    is first added to the unvetted cache then moved to the vetted
-	//    cache. This is a limitation of the inventory API and mimicks
-	//    how records are added and updated when using the politeiad
-	//    API.
-	//
-	// 3. Add the unvetted tokens to the inventory cache. They MUST be
-	//    added starting with the oldest timestamp.
+	// Get the partial record for all tokens. This also guarantees that all
+	// tokens being manipulated actually correspond to a record on the backend.
+	records := make(map[string]*backend.Record, len(allTokens))
+	for _, token := range allTokens {
+		r, err := t.tstore.RecordPartial(token, 0, nil, true)
+		if err != nil {
+			return err
+		}
+		records[r.RecordMetadata.Token] = r
+	}
+
+	// Sort records into vetted and unvetted groups.
+	vetted := make([]*backend.Record, len(allTokens))
+	unvetted := make([]*backend.Record, len(allTokens))
+	for _, token := range allTokens {
+		record := records[hex.EncodeToString(token)]
+		if record.RecordMetadata.State == backend.StateVetted {
+			vetted = append(vetted, record)
+		}
+		if record.RecordMetadata.State == backend.StateUnvetted {
+			unvetted = append(unvetted, record)
+		}
+	}
+
+	// Sort records from both groups by the timestamp of their record metadata,
+	// from oldest to newest. The order of the record inventory will be
+	// slightly different. On runtime, the timestamp order is through the most
+	// recent status change md. On this fsck rebuild, the order is through the
+	// record metadata's timestamp from their last update.
```

Comment:
Explain why, i.e. the record timestamp gets updated on both status changes and edits, so the status change timestamp is lost once a record is edited.